src/client/mod.rs

```diff
@@ -1,2 +1,68 @@
+pub mod db;
+pub mod coll;
+pub mod common;
 pub mod connstring;
 pub mod wire_protocol;
+
+use std::cell::{Cell, RefCell};
+use std::net::TcpStream;
+use std::sync::Arc;
+
+use client::db::Database;
+use client::connstring::ConnectionString;
+
+/// Interfaces with a MongoDB server or replica set.
+pub struct MongoClient {
+    req_id: Cell<i32>,
+    socket: Arc<RefCell<TcpStream>>,
+    config: ConnectionString,
+}
+
+impl MongoClient {
+    /// Creates a new MongoClient connected to a single MongoDB server.
+    pub fn new(host: &str, port: u16) -> Result<MongoClient, String> {
+        let config = ConnectionString::new(host, port);
+        MongoClient::with_config(config)
+    }
+
+    /// Creates a new MongoClient connected to a server or replica set using 
+    /// a MongoDB connection string URI as defined by 
+    /// [the manual](http://docs.mongodb.org/manual/reference/connection-string/).
+    pub fn with_uri(uri: &str) -> Result<MongoClient, String> {
+        let config = try!(connstring::parse(uri));
+        MongoClient::with_config(config)
+    }
+
+    fn with_config(config: ConnectionString) -> Result<MongoClient, String> {
+        let socket = try!(MongoClient::connect(&config));
+        Ok(MongoClient {
+            req_id: Cell::new(0),
+            socket: Arc::new(RefCell::new(socket)),
+            config: config,
+        })
+    }
+
+    /// Returns a unique operational request id.
+    pub fn get_req_id(&self) -> i32 {
+        self.req_id.set(self.req_id.get() + 1);
```

Comment:
Oops, I was looking at the socket RefCell, which also needed fixing. This may be a race condition; looking into it.